File: assistant/src/tools.ts
import { execSync, spawn } from "node:child_process";
import {
  readFileSync,
  writeFileSync,
  existsSync,
  readdirSync,
  statSync,
} from "node:fs";
import { join, resolve, basename } from "node:path";
import { homedir, hostname, platform, arch, cpus, totalmem, freemem } from "node:os";
import type { ToolEntry, ToolResult, AssistantConfig } from "./types.js";

function ok(output: string): ToolResult {
  return { output };
}
function err(message: string): ToolResult {
  return { output: message, error: message, isError: true };
}

// ─── Bash Tool ─────────────────────────────────────────────────────────

const bashTool: ToolEntry = {
  definition: {
    name: "bash",
    description:
      "Execute a shell command and return stdout+stderr. Use for running programs, installing packages, git operations, and system tasks. Commands run in the user's default shell.",
    input_schema: {
      type: "object",
      properties: {
        command: {
          type: "string",
          description: "The shell command to execute",
        },
        cwd: {
          type: "string",
          description: "Working directory (default: home directory)",
        },
        timeout: {
          type: "number",
          description: "Timeout in milliseconds (default: 30000)",
        },
      },
      required: ["command"],
    },
  },
  handler: async (input) => {
    const command = input.command as string;
    const cwd = (input.cwd as string) || homedir();
    const timeout = (input.timeout as number) || 30000;

    try {
      const result = execSync(command, {
        cwd,
        timeout,
        maxBuffer: 1024 * 1024,
        encoding: "utf-8",
        stdio: ["pipe", "pipe", "pipe"],
        env: { ...process.env, HOME: homedir() },
      });
      return ok(result || "(no output)");
    } catch (e: unknown) {
      const error = e as { stdout?: string; stderr?: string; status?: number; message?: string };
      const stdout = error.stdout || "";
      const stderr = error.stderr || "";
      const exitCode = error.status ?? 1;
      return err(
        `Exit code ${exitCode}\n${stdout}${stderr ? "\nSTDERR:\n" + stderr : ""}`.trim(),
      );
    }
  },
};

// ─── File Tools ────────────────────────────────────────────────────────

const readFileTool: ToolEntry = {
  definition: {
    name: "read_file",
    description:
      "Read the contents of a file. Returns the full text content. For large files, use offset and limit to read a portion.",
    input_schema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Absolute or relative file path" },
        offset: { type: "number", description: "Line number to start from (1-based)" },
        limit: { type: "number", description: "Number of lines to read" },
      },
      required: ["path"],
    },
  },
  handler: async (input) => {
    const filePath = resolve(input.path as string);
    if (!existsSync(filePath)) return err(`File not found: ${filePath}`);
    try {
      const content = readFileSync(filePath, "utf-8");
      const offset = input.offset as number | undefined;
      const limit = input.limit as number | undefined;
      if (offset || limit) {
        const lines = content.split("\n");
        const start = Math.max(0, (offset ?? 1) - 1);
        const end = limit ? start + limit : lines.length;
        return ok(
          lines
            .slice(start, end)
            .map((l, i) => `${String(start + i + 1).padStart(6)}|${l}`)
            .join("\n"),
        );
      }
      if (content.length > 100_000) {
        return ok(content.slice(0, 100_000) + "\n... (truncated, file is " + content.length + " bytes)");
      }
      return ok(content);
    } catch (e: unknown) {
      return err(`Failed to read: ${(e as Error).message}`);
    }
  },
};

const writeFileTool: ToolEntry = {
  definition: {
    name: "write_file",
    description: "Write content to a file. Creates the file if it doesn't exist, overwrites if it does.",
    input_schema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Absolute or relative file path" },
        content: { type: "string", description: "Content to write" },
      },
      required: ["path", "content"],
    },
  },
  handler: async (input) => {
    const filePath = resolve(input.path as string);
    try {
      writeFileSync(filePath, input.content as string);
      return ok(`Written to ${filePath}`);
    } catch (e: unknown) {
      return err(`Failed to write: ${(e as Error).message}`);
    }
  },
};

const editFileTool: ToolEntry = {
  definition: {
    name: "edit_file",
    description:
      "Edit a file by replacing an exact string match with new text. The old_string must match exactly (including whitespace).",
    input_schema: {
      type: "object",
      properties: {
        path: { type: "string", description: "File path" },
        old_string: { type: "string", description: "Exact text to find and replace" },
        new_string: { type: "string", description: "Replacement text" },
        replace_all: { type: "boolean", description: "Replace all occurrences (default: false)" },
      },
      required: ["path", "old_string", "new_string"],
    },
  },
  handler: async (input) => {
    const filePath = resolve(input.path as string);
    if (!existsSync(filePath)) return err(`File not found: ${filePath}`);
    try {
      let content = readFileSync(filePath, "utf-8");
      const oldStr = input.old_string as string;
      const newStr = input.new_string as string;
      if (!content.includes(oldStr)) return err("old_string not found in file");
      if (input.replace_all) {
        content = content.split(oldStr).join(newStr);
      } else {
        const idx = content.indexOf(oldStr);
        content = content.slice(0, idx) + newStr + content.slice(idx + oldStr.length);
      }
      writeFileSync(filePath, content);
      return ok(`Edited ${filePath}`);
    } catch (e: unknown) {
      return err(`Failed to edit: ${(e as Error).message}`);
    }
  },
};

const listFilesTool: ToolEntry = {
  definition: {
    name: "list_files",
    description: "List files and directories at a given path. Shows names, sizes, and types.",
    input_schema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Directory path to list" },
        recursive: { type: "boolean", description: "List recursively (default: false, max 200 entries)" },
        pattern: { type: "string", description: "Glob pattern to filter (e.g. '*.ts')" },
      },
      required: ["path"],
    },
  },
  handler: async (input) => {
    const dirPath = resolve(input.path as string);
    if (!existsSync(dirPath)) return err(`Path not found: ${dirPath}`);

    if (input.pattern) {
      try {
        const result = execSync(
          `find ${JSON.stringify(dirPath)} -maxdepth ${input.recursive ? 5 : 1} -name ${JSON.stringify(input.pattern)} 2>/dev/null | head -200`,
          { encoding: "utf-8", timeout: 5000 },
        );
        return ok(result || "(no matches)");
      } catch {
        return ok("(no matches)");
      }
    }

    try {
      const entries = readdirSync(dirPath, { withFileTypes: true });
      const lines = entries.slice(0, 200).map((e) => {
        const full = join(dirPath, e.name);
        if (e.isDirectory()) return `  ${e.name}/`;
        try {
          const s = statSync(full);
          const size = s.size < 1024 ? `${s.size}B` : s.size < 1048576 ? `${(s.size / 1024).toFixed(0)}K` : `${(s.size / 1048576).toFixed(1)}M`;
          return `  ${e.name}  (${size})`;
        } catch {
          return `  ${e.name}`;
        }
      });
      return ok(lines.join("\n"));
    } catch (e: unknown) {
      return err(`Failed to list: ${(e as Error).message}`);
    }
  },
};

const grepTool: ToolEntry = {
  definition: {
    name: "grep",
    description: "Search for a pattern in files using ripgrep. Returns matching lines with file paths and line numbers.",
    input_schema: {
      type: "object",
      properties: {
        pattern: { type: "string", description: "Regex pattern to search for" },
        path: { type: "string", description: "Directory or file to search in" },
        glob: { type: "string", description: "File glob filter (e.g. '*.ts')" },
        case_insensitive: { type: "boolean", description: "Case insensitive search" },
      },
      required: ["pattern", "path"],
    },
  },
  handler: async (input) => {
    const args = ["rg", "--max-count=50", "--line-number"];
    if (input.case_insensitive) args.push("-i");
    if (input.glob) args.push("--glob", input.glob as string);
    args.push(input.pattern as string, resolve(input.path as string));

    try {
      const result = execSync(args.join(" "), {
        encoding: "utf-8",
        timeout: 10000,
        maxBuffer: 512 * 1024,
      });
      return ok(result || "(no matches)");
    } catch (e: unknown) {
      const error = e as { stdout?: string; status?: number };
      if (error.status === 1) return ok("(no matches)");
      return err(error.stdout || "Search failed");
    }
  },
};

// ─── Web Tool ──────────────────────────────────────────────────────────

const webFetchTool: ToolEntry = {
  definition: {
    name: "web_fetch",
    description: "Fetch content from a URL and return it as text. Useful for checking APIs, downloading data, or reading web pages.",
    input_schema: {
      type: "object",
      properties: {
        url: { type: "string", description: "URL to fetch" },
        method: { type: "string", description: "HTTP method (default: GET)" },
        headers: {
          type: "object",
          description: "Request headers",
          additionalProperties: { type: "string" },
        },
        body: { type: "string", description: "Request body (for POST/PUT)" },
      },
      required: ["url"],
    },
  },
  handler: async (input) => {
    const url = input.url as string;
    const method = (input.method as string) || "GET";
    try {
      const opts: RequestInit = {
        method,
        headers: (input.headers as Record<string, string>) || {},
        signal: AbortSignal.timeout(15000),
      };
      if (input.body) opts.body = input.body as string;
      const res = await fetch(url, opts);
      const text = await res.text();
      const truncated = text.length > 50_000 ? text.slice(0, 50_000) + "\n...(truncated)" : text;
      return ok(`HTTP ${res.status} ${res.statusText}\n\n${truncated}`);
    } catch (e: unknown) {
      return err(`Fetch failed: ${(e as Error).message}`);
    }
  },
};

// ─── System Tool ───────────────────────────────────────────────────────

const systemInfoTool: ToolEntry = {
  definition: {
    name: "system_info",
    description:
      "Get system information: hostname, OS, CPU, memory, disk, uptime, and running processes.",
    input_schema: {
      type: "object",
      properties: {
        section: {
          type: "string",
          enum: ["overview", "processes", "disk", "network"],
          description: "Which info section (default: overview)",
        },
      },
    },
  },
  handler: async (input) => {
    const section = (input.section as string) || "overview";
    switch (section) {
      case "processes":
        try {
          const ps = execSync("ps aux --sort=-%mem 2>/dev/null | head -20 || ps aux | head -20", {
            encoding: "utf-8",
            timeout: 5000,
          });
          return ok(ps);
        } catch {
          return ok("(could not list processes)");
        }
      case "disk":
        try {
          return ok(execSync("df -h", { encoding: "utf-8", timeout: 5000 }));
        } catch {
          return ok("(could not get disk info)");
        }
      case "network":
        try {
          const ifconfig = execSync(
            "ifconfig 2>/dev/null | grep -E 'inet |flags' | head -20 || ip addr show 2>/dev/null | head -20",
            { encoding: "utf-8", timeout: 5000 },
          );
          return ok(ifconfig);
        } catch {
          return ok("(could not get network info)");
        }
      default: {
        const totalMem = (totalmem() / 1073741824).toFixed(1);
        const freeMem = (freemem() / 1073741824).toFixed(1);
        const upSec = process.uptime();
        const info = [
          `Hostname: ${hostname()}`,
          `Platform: ${platform()} ${arch()}`,
          `CPUs:     ${cpus().length}x ${cpus()[0]?.model || "unknown"}`,
          `Memory:   ${freeMem} GB free / ${totalMem} GB total`,
          `Node:     ${process.version}`,
          `Uptime:   ${Math.floor(upSec / 3600)}h ${Math.floor((upSec % 3600) / 60)}m`,
          `User:     ${homedir().split("/").pop()}`,
          `CWD:      ${process.cwd()}`,
        ];
        return ok(info.join("\n"));
      }
    }
  },
};

// ─── Registry ──────────────────────────────────────────────────────────

export function buildToolRegistry(config: AssistantConfig): Map<string, ToolEntry> {
  const registry = new Map<string, ToolEntry>();

  if (config.tools.bash.enabled) {
    registry.set("bash", bashTool);
  }
  if (config.tools.files.enabled) {
    registry.set("read_file", readFileTool);
    registry.set("write_file", writeFileTool);
    registry.set("edit_file", editFileTool);
    registry.set("list_files", listFilesTool);
    registry.set("grep", grepTool);
  }
  if (config.tools.web.enabled) {
    registry.set("web_fetch", webFetchTool);
  }
  if (config.tools.system.enabled) {
    registry.set("system_info", systemInfoTool);
  }

  return registry;
}

File: scripts/_agent_engine/tools.py
"""Tool definitions and async execution for the Claude tool_use protocol."""

import asyncio
import logging
import os

from .config import DANGEROUS_PATTERNS, DANGEROUS_PREFIXES

log = logging.getLogger("kmac-agent")

TOOLS = [
    {
        "name": "bash",
        "description": (
            "Execute a shell command and return stdout+stderr. Use for "
            "running programs, git, system commands, package management, "
            "or any terminal task."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "command": {"type": "string", "description": "Shell command"}
            },
            "required": ["command"],
        },
    },
    {
        "name": "read_file",
        "description": "Read a file's contents with line numbers.",
        "input_schema": {
            "type": "object",
            "properties": {
                "path": {"type": "string", "description": "File path"},
                "offset": {
                    "type": "integer",
                    "description": "Start line (1-indexed)",
                },
                "limit": {
                    "type": "integer",
                    "description": "Max lines to return",
                },
            },
            "required": ["path"],
        },
    },
    {
        "name": "write_file",
        "description": "Create or overwrite a file.",
        "input_schema": {
            "type": "object",
            "properties": {
                "path": {"type": "string", "description": "File path"},
                "content": {"type": "string", "description": "Full content"},
            },
            "required": ["path", "content"],
        },
    },
    {
        "name": "edit_file",
        "description": (
            "Find and replace an exact string in a file. old_string must "
            "be unique. Always read_file first to see current content."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "path": {"type": "string", "description": "File to edit"},
                "old_string": {"type": "string", "description": "Exact text to find"},
                "new_string": {"type": "string", "description": "Replacement text"},
            },
            "required": ["path", "old_string", "new_string"],
        },
    },
    {
        "name": "list_dir",
        "description": "List files and directories at a path.",
        "input_schema": {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Directory (default: cwd)",
                },
                "depth": {
                    "type": "integer",
                    "description": "Max depth (default: 2)",
                },
            },
        },
    },
    {
        "name": "grep_search",
        "description": (
            "Search for a regex pattern across files. Returns matching "
            "lines with paths and line numbers. Uses ripgrep if available."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "pattern": {"type": "string", "description": "Regex pattern"},
                "path": {
                    "type": "string",
                    "description": "Search directory (default: cwd)",
                },
                "include": {
                    "type": "string",
                    "description": "File glob, e.g. '*.py'",
                },
            },
            "required": ["pattern"],
        },
    },
]


def _trunc(text: str, limit: int = 80000) -> str:
    if len(text) <= limit:
        return text
    half = limit // 2
    cut = len(text) - limit
    return text[:half] + f"\n\n... ({cut} chars truncated) ...\n\n" + text[-half:]


async def execute(name: str, inp: dict, timeout: int = 120):
    """Execute a tool. Returns (result_text, display_preview)."""
    try:
        if name == "bash":
            return await _bash(inp, timeout)
        if name == "read_file":
            return _read_file(inp)
        if name == "write_file":
            return _write_file(inp)
        if name == "edit_file":
            return _edit_file(inp)
        if name == "list_dir":
            return await _list_dir(inp)
        if name == "grep_search":
            return await _grep(inp)
        return f"Unknown tool: {name}", f"Unknown tool: {name}"
    except Exception as e:
        msg = f"Error ({name}): {e}"
        return msg, msg


def _check_dangerous(cmd: str) -> str | None:
    """Return a warning message if the command looks dangerous, else None."""
    lower = cmd.lower().strip()
    for pat in DANGEROUS_PATTERNS:
        if pat.lower() in lower:
            return f"BLOCKED: dangerous command detected ({pat})"
    for prefix in DANGEROUS_PREFIXES:
        if lower.startswith(prefix.lower()):
            return f"BLOCKED: dangerous command prefix ({prefix})"
    return None


async def _bash(inp, timeout):
    cmd = inp["command"]
    warning = _check_dangerous(cmd)
    if warning:
        log.warning("Blocked dangerous command: %s", cmd[:100])
        return warning, warning
    try:
        proc = await asyncio.create_subprocess_shell(
            cmd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await asyncio.wait_for(
            proc.communicate(), timeout=timeout
        )
        out = (stdout or b"").decode("utf-8", errors="replace")
        err = (stderr or b"").decode("utf-8", errors="replace")
        full = out + err
        if not full.strip():
            full = f"(exit code {proc.returncode})"
        lines = full.rstrip("\n").split("\n")
        preview = "\n".join(lines[:20])
        if len(lines) > 20:
            preview += f"\n... ({len(lines) - 20} more lines)"
        return _trunc(full), preview
    except asyncio.TimeoutError:
        return "Error: command timed out", "Error: timed out"


def _read_file(inp):
    path = inp["path"]
    with open(path, "r") as f:
        all_lines = f.readlines()
    offset = max(0, inp.get("offset", 1) - 1)
    limit = inp.get("limit", len(all_lines))
    sel = all_lines[offset : offset + limit]
    numbered = "".join(
        f"{offset + i + 1:>6}|{ln}" for i, ln in enumerate(sel)
    )
    return _trunc(numbered), f"{len(sel)} of {len(all_lines)} lines"


def _write_file(inp):
    path, content = inp["path"], inp["content"]
    parent = os.path.dirname(path)
    if parent:
        os.makedirs(parent, exist_ok=True)
    tmp = path + ".tmp"
    with open(tmp, "w") as f:
        f.write(content)
    os.replace(tmp, path)
    lc = content.count("\n") + (
        1 if content and not content.endswith("\n") else 0
    )
    msg = f"Wrote {lc} lines to {path}"
    return msg, msg


def _edit_file(inp):
    path, old, new = inp["path"], inp["old_string"], inp["new_string"]
    with open(path, "r") as f:
        content = f.read()
    n = content.count(old)
    if n == 0:
        return (
            f"Error: old_string not found in {path}",
            "old_string not found",
        )
    if n > 1:
        return (
            f"Error: old_string found {n} times — must be unique",
            f"found {n} times",
        )
    content = content.replace(old, new, 1)
    tmp = path + ".tmp"
    with open(tmp, "w") as f:
        f.write(content)
    os.replace(tmp, path)
    return f"Edited {path}", f"edited {path}"


async def _list_dir(inp):
    path = inp.get("path", ".")
    depth = inp.get("depth", 2)
    cmd = (
        f"find '{path}' -maxdepth {depth} -not -path '*/.*' "
        "2>/dev/null | sort | head -100"
    )
    proc = await asyncio.create_subprocess_shell(
        cmd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=10)
    out = stdout.decode().strip()
    return out or "(empty)", out[:500] if out else "(empty)"


async def _grep(inp):
    pattern, path = inp["pattern"], inp.get("path", ".")
    include = inp.get("include", "")
    rg_check = await asyncio.create_subprocess_shell(
        "command -v rg",
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    await rg_check.communicate()
    if rg_check.returncode == 0:
        cmd = f"rg -n --max-count 50 --no-heading"
        if include:
            cmd += f" -g '{include}'"
        cmd += f" '{pattern}' '{path}'"
    else:
        cmd = f"grep -rn"
        if include:
            cmd += f" --include='{include}'"
        cmd += f" '{pattern}' '{path}' | head -50"
    proc = await asyncio.create_subprocess_shell(
        cmd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=30)
    out = stdout.decode().strip()
    lines = out.split("\n") if out else []
    preview = "\n".join(lines[:20])
    if len(lines) > 20:
        preview += f"\n... ({len(lines) - 20} more)"
    return _trunc(out) if out else "No matches found.", preview or "No matches."
